astropy/utils/data.py

```diff
@@ -1311,59 +1411,332 @@ def _get_download_cache_locs():
+                hexdigest_file = compute_hash(h)
+                if hexdigest_file != hexdigest:
+                    bad_hashes[u] = h
+        for h in leftover_files:
+            h_base = os.path.basename(h)
+            if len(h_base) >= 32 and re.match(r"[0-9a-f]+", h_base):
+                abandoned_files.add(h)
+        leftover_files -= abandoned_files
+
+        msgs = []
+        if nonexistent_targets:
+            msgs.append(
+                f"URL(s) point(s) to nonexistent file(s): {nonexistent_targets}")
+        if bad_locations:
+            msgs.append(
+                f"URL(s) point(s) to file(s) outside {dldir}: {bad_locations}")
+        if bad_hashes:
+            msgs.append(
+                f"Filename(s) does not match hash(es) of contents: {bad_hashes}")
+        if abandoned_files:
+            msgs.append(
+                f"Apparently abandoned files: {abandoned_files}")
+        if msgs:
+            raise CacheDamaged(
+                "\n".join(msgs),
+                bad_urls=(list(nonexistent_targets.keys())
+                          + list(bad_locations.keys())),
+                bad_files=(list(bad_hashes.values())
+                           + list(abandoned_files)))
         else:
-            return
-    msg = ("Unable to acquire lock for cache directory ({0} exists). "
-           "You may need to delete the lock if the python interpreter wasn't "
-           "shut down properly.")
-    raise RuntimeError(msg.format(lockdir))
+            return leftover_files
 
 
-def _release_download_cache_lock():
-    lockdir = os.path.join(_get_download_cache_locs()[0], 'lock')
+def import_file_to_cache(url_key, filename,
+                         hexdigest=None,
+                         remove_original=False):
+    """Import the on-disk file specified by filename to the cache.
 
-    if os.path.isdir(lockdir):
-        # if the pid file is present, be sure to remove it
-        pidfn = os.path.join(lockdir, 'pid')
-        if os.path.exists(pidfn):
-            os.remove(pidfn)
-        os.rmdir(lockdir)
-    else:
-        msg = 'Error releasing lock. "{0}" either does not exist or is not ' +\
-              'a directory.'
-        raise RuntimeError(msg.format(lockdir))
+    Parameters
+    ----------
+    url_key : str
+        The key to index the file under. This should probably be
+        the URL where the file was located, though if you obtained
+        it from a mirror you should use the URL of the primary
+        location.
+    filename : str
+        The file whose contents you want to import.
+    hexdigest : str, optional
+        The cryptographic hash of the file, as computed by
+        `compute_hash`. If it is not provided, `compute_hash`
+        will be called. This argument is available in case
+        it is easier to compute the hash progressively as
+        the file is downloaded.
+    remove_original : bool
+        Whether to remove the original file once import is complete.
```

Comment:
A clarification I find useful when I try to use this function.
```suggestion
        Whether to remove the original file (``filename``) once import is complete.
```